test/distributed/test_tiled_mlp_distributed.py

```diff
@@ -0,0 +1,348 @@
+        else:  # swiglu
+            tiled_mlp = LigerTiledSwiGLUMLP(config=config, num_shards=num_shards).to(device).to(dtype)
+
+        tiled_mlp.gate_proj.weight.data = G
+        tiled_mlp.up_proj.weight.data = U
+        tiled_mlp.down_proj.weight.data = D
+
+        # Wrap with DDP
+        ddp_mlp = DDP(tiled_mlp, device_ids=[rank])
+
+        # Forward pass
+        output = ddp_mlp(x)
+
+        # Backward pass with same gradient across all ranks
+        torch.manual_seed(42)  # Same gradient for all ranks
+        grad_output = torch.randn_like(output)
+        output.backward(grad_output)
+
+        # Verify that module is detected as DDP
+        assert hasattr(ddp_mlp.module, "gate_proj"), "Model structure is correct"
+
+        # Verify gradients exist
+        assert ddp_mlp.module.gate_proj.weight.grad is not None
+        assert ddp_mlp.module.up_proj.weight.grad is not None
+        assert ddp_mlp.module.down_proj.weight.grad is not None
+
+        # Verify gradient synchronization across ranks
+        # All ranks should have identical gradients after DDP synchronization
+        gate_grad = ddp_mlp.module.gate_proj.weight.grad.clone()
+        up_grad = ddp_mlp.module.up_proj.weight.grad.clone()
+        down_grad = ddp_mlp.module.down_proj.weight.grad.clone()
+
+        # Gather gradients from all ranks to rank 0
+        if rank == 0:
+            gate_grads = [torch.zeros_like(gate_grad) for _ in range(world_size)]
+            up_grads = [torch.zeros_like(up_grad) for _ in range(world_size)]
+            down_grads = [torch.zeros_like(down_grad) for _ in range(world_size)]
+        else:
+            gate_grads = None
+            up_grads = None
+            down_grads = None
+
+        dist.gather(gate_grad, gate_grads, dst=0)
+        dist.gather(up_grad, up_grads, dst=0)
+        dist.gather(down_grad, down_grads, dst=0)
+
+        # Rank 0 verifies all gradients are synchronized
+        if rank == 0:
+            for i in range(1, world_size):
+                torch.testing.assert_close(
+                    gate_grads[0],
+                    gate_grads[i],
+                    rtol=1e-5,
+                    atol=1e-5,
+                    msg=f"Gate gradients not synchronized between rank 0 and rank {i}",
+                )
+                torch.testing.assert_close(
+                    up_grads[0],
+                    up_grads[i],
+                    rtol=1e-5,
+                    atol=1e-5,
+                    msg=f"Up gradients not synchronized between rank 0 and rank {i}",
+                )
+                torch.testing.assert_close(
+                    down_grads[0],
+                    down_grads[i],
+                    rtol=1e-5,
+                    atol=1e-5,
+                    msg=f"Down gradients not synchronized between rank 0 and rank {i}",
+                )
+
+        # Barrier to ensure all ranks complete
+        dist.barrier()
+
+    finally:
+        cleanup_distributed()
```

Comment:
You have to move `barrier()` below `finally` in case some ranks occur error so that they never reach this barrier and hang indefinitely.
```suggestion
    finally:
    	# Barrier to ensure all ranks complete
        dist.barrier()
        cleanup_distributed()
```